scheduler/src/cook/kubernetes/controller.clj

```diff
@@ -0,0 +1,140 @@
+    (swap! update map-atom key value)))
+
+;; REVIEW: Review handle-status-update
+(defn pod-has-just-completed
+  "A pod has completed."
+  [{:keys [synthesized-state pod] :as existing-state-dictionary}]
+  (case synthesized-state
+    :existing/failed
+    (do
+      ; TODO: Extract failure reason, etc and store in datomic, via refactored handle-status-update.
+      {:expected-state :expected/completed})
+    :existing/succeeded
+    (do
+      ; TODO: Mark job as success in datomic, via refactored handle-status-update.
+      {:expected-state :expected/completed})))
+
+(defn process
+  "Visit this pod-name, processing the new level-state. Returns the new expected state. Returns
+  empty dictionary to indicate that the result should be deleted. NOTE: Must be invoked with the lock."
+  [{:keys [api-client existing-state-map expected-state-map] :as kcc} pod-name]
+  (let [{:keys [expected-state] :as expected-state-dict} (get @expected-state-map pod-name)
+        {:keys [synthesized-existing-state pod] :as existing-state-dict} (get @existing-state-map pod-name)
+        ; .toString on a pod is incredibly large. Make a version thats been elided.
+        expected-state-dict-elided (if (:launch-pod expected-state-dict)
+                                (assoc expected-state-dict :launch-pod [:elided-for-brevity])
+                                expected-state-dict)]
+    (log/info "Processing " pod-name ": ((" expected-state-dict-elided " ===== " existing-state-dict "))")
+    (let
+        [new-expected-state (case (vector (or expected-state :missing) (or synthesized-existing-state :missing))
+                             [:expected/running :missing] (launch-task api-client expected-state-dict)
+                             [:expected/running :existing/running] expected-state-dict ; TODO: Need to call handle-status-update to move the task to running state.
+                             [:expected/running :existing/succeeded] (pod-has-just-completed existing-state-dict)
+                             [:expected/completed :existing/succeeded] (remove-finalization-if-set api-client expected-state-dict pod)
+                             [:expected/running :existing/failed] (remove-finalization-if-set api-client expected-state-dict pod)
+                             [:expected/completed :existing/failed] (remove-finalization-if-set api-client expected-state-dict pod)
+                             [:expected/completed :missing] nil ; Cause it to be deleted.
+                             [:expected/killed :existing/running] (kill-task api-client expected-state-dict pod) ; TODO: Where does the datomic update occur? Do we do it when we do [expected/killed :existing/failed], to be similar to mesos, we update it only when the backend says its dead?
+
+                             [:expected/killed :existing/succeeded] (remove-finalization-if-set api-client expected-state-dict pod)
+                             [:expected/killed :existing/failed]
```

Comment:
I would think in this case it would make more sense to call `pod-completed`, then transition to `expected/failed - existing/failed` which will delete the pod.

Reply:
Maybe. Not sure. I think it depends on exactly how the handle-status-update works for killed/failed races. In this case, you already killed it, so is there any update to do? Is the update special, or is it something done in the normal handle-status-update logic?